Add functions to calculate truth

Bai-Li-NOAA opened the pull request:
<!---
Thanks for opening a PR. This commented text will **NOT** appear in the final PR. Toggle between Write and Preview to see what your PR will look like without the comments.
-->

# What is the feature?
* Add R functions to calculate "truth" for a stock using data from {ecosystemdata}. The "truth" includes "true" index and age composition data for biomass, catch, numbers, and weight. These datasets are summarized at both monthly and yearly resolutions. 

# How have you implemented the solution?
* Add functions to calculate "truth" for a stock and add tests
* Add package-level documentation using `usethis::use_package_doc()`
* Add {testthat} template for testing the new functions
* Update `DESCRIPTION` to reflect new dependencies
* Add editor rulers in VSCode

# Does the PR impact any other area of the project, maybe another repo?
*

Bai-Li-NOAA committed:
feat(calc_truth): Add functions to calculate truth

* Add functions to calculate "truth" for a stock and add tests
* Add package-level documentation using `usethis::use_package_doc()`
* Add {testthat} template for testing the new functions
* Update `DESCRIPTION` to reflect new dependencies
* Add editor rulers in VSCode

Bai-Li-NOAA commented on R/use_template.R:
I copied `R/use_template.R` from the FIMS repo, but I’m now thinking that might not be the right approach. Would it make more sense to import `use_template()` from FIMS in `ecosystemom-package.R` instead? Are we able to use the template directly from `FIMS/inst/templates`? Another option could be to make `use_template.R` a standalone file that can be reused across repos. See [`usethis::use_standalone()`](https://usethis.r-lib.org/reference/use_standalone.html).

Bai-Li-NOAA commented:
Make sure we use species and group as column names in the output tibble.

kellijohnson-NOAA commented on DESCRIPTION:
```diff
@@ -1,19 +1,27 @@
 Package: ecosystemom
-Title: What the Package Does (One Line, Title Case)
+Title: Sample data from Ecopath with Ecosim data sets
```
```suggestion
Title: Create sampled data from standardized ecosystem output
```

Bai-Li-NOAA replied:
The suggestion has been accepted.

kellijohnson-NOAA commented on LICENSE.md:
Should this file have the .md extension or just be LICENSE like the original file?

Bai-Li-NOAA replied:
Great question. I used `usethis::use_gpl_license()`, which adds `GPL-3` to the `DESCRIPTION`, creates a `LICENSE.md`, and adds it to `.Rbuildignore`. Since CRAN already includes the GPL license automatically, we don’t need to include the full text in a `LICENSE` file and `.Rbuildignore` excludes `LICENSE.md` in the CRAN submission. For more details, see the "[Key files](https://r-pkgs.org/license.html#key-files)" section of R Packages (2e). 
If we want to specify the year and copyright holder, we can manually create a LICENSE file. See how it's done in the [`dplyr` package](https://github.com/tidyverse/dplyr/blob/main/LICENSE).

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' Calculate "truth" for a stock using {ecosystemdata} outputs
```
```suggestion
#' Calculate "truth" for a species using {ecosystemdata} outputs
```

Bai-Li-NOAA replied:
Done. I also replaced 'stock' with 'species' throughout the rest of the package.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' 
+#' @description This function generates standardized "truth" outputs using data 
+#' from {ecosystemdata}. The "truth" include "true" index and age composition
+#' (both monthly and yearly) for a specified stock.
```
```suggestion
#' @description The truth from an ecosystem model is often in different units
#' than what is needed to sample from. This function takes an {ecosystemdata}
#' data set and calculates additional true information such as converting
#' monthly age-composition information into annual data. No sampling is
#' performed just mathematical operations.
```

Bai-Li-NOAA replied:
The suggestion has been accepted.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' 
+#' @details Currently, this function supports calculating "truth" for biomass, 
+#' catch, numbers, and weight. Numbers-at-age are calculated by dividing 
+#' biomass-at-age by weight-at-age.
```
```suggestion
#' @details The following values are calculated within this function:
#'   * biomass
#'   * catch
#'   * numbers at age by dividing biomass-at-age by weight-at-age information
#'   * weight
```

Bai-Li-NOAA replied:
The suggestion has been accepted.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' biomass-at-age by weight-at-age.
+#' 
+#' @param data A tibble containing ecosystem model (e.g., Ecopath with Ecosim) 
+#' outputs from {ecosystemdata}.
```
Is there a specific function in ecosystemdata that we can reference here, as in "outputs from [ecosystemdata::x()]"?

Bai-Li-NOAA replied:
Yes. I have changed it to "outputs from [ecosystemdata::load_model()]"

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' 
+#' @param data A tibble containing ecosystem model (e.g., Ecopath with Ecosim) 
+#' outputs from {ecosystemdata}.
+#' @param stock_name A character string specifying the stock name to subset.
```
```suggestion
#' @param species_name A string specifying the species name to subset.
```

Bai-Li-NOAA replied:
Done. I also replaced 'character string' with 'string' throughout the rest of the package.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' 
+#' @return A tibble with nested "truth" (e.g., biomass, catch, 
+#' numbers, or weight), categorized by type (index or age composition) and 
+#' timestep (monthly or yearly).
```
```suggestion
#' @return
#' A nested tibble where each row holds an output type (e.g., biomass, catch, 
#' numbers, or weight) for a given timestep category (e.g., monthly or yearly).
#' Thus, data within a row in returned output can be expanded to reveal the
#' full data, e.g., all months for an output type.
```

Bai-Li-NOAA replied:
The suggestion has been accepted.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' data <- ewe_nwatlantic_base
+#' truth <- calc_truth(
+#'   data,
+#'   stock_name = "menhaden"
```
```suggestion
#'   species_name = "menhaden"
```

kellijohnson-NOAA replied:
I stopped noting these changes for species, stock, population because it will probably be easier to just do a search and replace.

Bai-Li-NOAA replied:
Done.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+  missing_data <- setdiff(labels, unique(data[["type"]]))
+  if (length(missing_data) > 0) {
+    cli::cli_abort(c(
+      "{.val {missing_data}} at age data is missing for {stock_name}.",
```
```suggestion
     "x" = "{.val {missing_data}}-at-age data is missing for {stock_name}.",
```

Bai-Li-NOAA replied:
Done

kellijohnson-NOAA reviewed (changes requested):
@Bai-Li-NOAA I only got part way through my review, mainly comments on documentation. Perhaps if you have time before I can pick this review back up you can integrate the suggested changes and the changes from copilot and then I will finish?

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+  if (!stock_name %in% unique(data[["species"]])) {
+    cli::cli_abort(c(
+      "Stock {.val {stock_name}} not found in the data.",
+      "Please check the stock name and try again."
```
```suggestion
      "x" = "Stock {.val {stock_name}} not found in the data.",
      "i" = "Please check the stock name and try again."
```
You could list the available ones or ones that start with the same letter in the "information" that is returned on line 35.

Bai-Li-NOAA replied:
Done. I have listed the available ones.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+  if (length(missing_data) > 0) {
+    cli::cli_abort(c(
+      "{.val {missing_data}} at age data is missing for {stock_name}.",
+      "Please ensure that the data contains {missing_data} information."
```
```suggestion
      "i" = "Please ensure that the data contains {missing_data} information."
```
You could provide information on the data types that are available.

Bai-Li-NOAA replied:
Done.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+  # Extract monthly data by labels
+  truth_monthly <- purrr::map(
+    labels,
+    ~ calc_truth_monthly(data, stock_name, .x)
```
The recommendations for purrr suggest
* An anonymous function, e.g. \(x) x + 1 or function(x) x + 1.
* A formula, e.g. ~ .x + 1. You must use .x to refer to the first argument. Only recommended if you require backward compatibility with older versions of R.
Where, the latter is what you are using here but it is only suggested for backward compatibility.

Bai-Li-NOAA replied:
Done.

kellijohnson-NOAA commented on R/sample_lognormal.R:
```diff
@@ -40,14 +40,14 @@
 #' # because a single random number is generated for the entire vector.
 #' set.seed(123)
-#' sample_lognormal(1:10, sd = 1)
+#' ecosystemom:::sample_lognormal(1:10, sd = 1)
```
I think that we just want to export the sampling functions instead.

Bai-Li-NOAA replied:
Done

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' truth <- calc_truth(
+#'   data,
+#'   stock_name = "menhaden"
+#' )
```
This example should be extended to also include how to unnest the data, e.g.,
```
truth_long <- truth |>
  tidyr::unnest(truth_om)
```

Bai-Li-NOAA replied:
Done.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#'   stock_name = "menhaden"
+#' )
+#' @export
+#' @rdname calc_truth
```
Do we need this line because the rd file is named the same as the function?

Bai-Li-NOAA replied:
No, we don't need this line and I have removed it from the file.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' 
+#' @return A tibble with nested "truth" (e.g., biomass, catch, 
+#' numbers, or weight), categorized by type (index or age composition) and 
+#' timestep (monthly or yearly).
```
"timestep" is actually "time step" or "time-step" depending on the context. So, we should either change this to "truth_time_step" or "truth_interval" or something that is just one word. The changes will need to be made throughout and not just right here. Linters have a problem with "truth_timestep".

Bai-Li-NOAA replied:
changed "timestep" to "time_step".

kellijohnson-NOAA commented on tests/testthat/test-calc_truth_biomass.R:
```diff
@@ -0,0 +1,164 @@
+  expect_equal(
+    object = truth[["truth_timestep"]] |> unique(),
+    expected = expected_truth_timestep
+  )
```
nit: the calls to unique should either be on their own line or do not use the pipe and just use standard `unique(truth["*"]])` syntax.

Bai-Li-NOAA replied:
Done

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,328 @@
+#' 
+#' @param data A tibble containing ecosystem model (e.g., Ecopath with Ecosim) 
+#' outputs from {ecosystemdata}.
+#' @param stock_name A character string specifying the stock name to subset.
```
`stock_name` or `species_name` implies to me that you are only allowed to put one entry, i.e., that it will not accept a vector. I do not know if we need to rename this argument or just provide more details/examples indicating that you can pull information for multiple species at a time.

Bai-Li-NOAA replied:
I'll leave this comment open for now. Currently, `calculate_truth()` only accepts a single entry. Users can use functions from other packages (e.g., {purrr}) to achieve the desired result. I’ve added an example to the documentation for now, but let me know if you think it would be better to expand the function to accept a vector directly.

kellijohnson-NOAA commented:
@Bai-Li-NOAA would you mind updating this PR with some of the name changes that I have suggested if you agree with them. I think that we have talked about them in the past. Then I will finish reviewing the few pieces that are left this week?

Bai-Li-NOAA commented on tests/testthat/test-calc_truth_biomass.R:
```diff
@@ -0,0 +1,164 @@
+    (n_label_index * n_timestep) +
+    (n_label_agecomp * n_timestep)
+  )
+  #' @description Test that combined truth from multiple stocks has the expected number of rows..
```
Done

Bai-Li-NOAA commented on tests/testthat/test-calc_truth_biomass.R:
```diff
@@ -0,0 +1,164 @@
+  )
+})
+
+test_that("calc_truth() returns correct outputs for edge cases", {
```
Done.

Bai-Li-NOAA commented on R/use_template.R:
```diff
@@ -0,0 +1,59 @@
+      template = "testthat_template.R",
+      save_as = path,
+      data = list(function_name = function_name),
+      package = "ecosystemdata"
```
Done.

Bai-Li-NOAA commented on tests/testthat/test-calc_truth_biomass.R:
```diff
@@ -0,0 +1,164 @@
+    ) |>
+    dplyr::arrange(truth_label, truth_type, year, group)
+
+  #' @description Test that yearly output from calc_truthmatches the averaged monthly output.
```
Done.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+    purrr::set_names(names(truth_values))
+
+  # Combine all "truth" data into a single tibble
+  truth <- dplyr::tibble(
```
I think that by assigning the last item to an object the output is returned invisibly rather than printed to the screen. I am not sure if that is what you intended here?

Bai-Li-NOAA replied:
Thanks, I've updated all functions and avoided assigning the last item to an object.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+#' A nested tibble where each row holds an output type (e.g., biomass, catch, 
+#' numbers, or weight) for a given time step category (e.g., monthly or yearly).
+#' Thus, data within a row in returned output can be expanded to reveal the
+#' full data, e.g., all months for an output type.
```
I think this section needs more details, for example the structure of the nested data frame should be given as a described list and then the unnested list should also be described column by column. For example, there are no details on what an NA value for group means and I expected to be able to find that information here.

Bai-Li-NOAA replied:
Thanks, I've added more detailed descriptions for all columns.

kellijohnson-NOAA commented on tests/testthat/test-calc_truth_biomass.R:
```diff
@@ -0,0 +1,158 @@
+})
+
+## Edge handling ----
+test_that("calc_truth() output can be binded", {
```
Spelling mistake in "binded" where it should be "bound"

Bai-Li-NOAA replied:
Done, and thanks for noticing!

kellijohnson-NOAA commented on tests/testthat/test-calc_truth_biomass.R:
```diff
@@ -0,0 +1,158 @@
+})
+
+test_that("calc_truth() returns correct outputs for edge cases", {
+  #' @description Test that calc_truth() returns NAs for 'group' when the species does not have multiple groups.
```
Should the group type be NA or should it be the range that is included, i.e., 0-10+.

Bai-Li-NOAA replied:
Perhaps use NA for now, unless the user provides additional information in the {ecosystemdata} about the included range.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+#' 
+#' @param data A tibble containing ecosystem model (e.g., Ecopath with Ecosim) 
+#' outputs from [ecosystemdata::load_model()].
+#' @param species_name A string specifying the species name to subset.
```
```suggestion
#' @param species_name A string specifying the desired species you want to summarize information for. This string must match an entry in `data[["species"]]` and it currently cannot be a vector, just a single string. See the examples section for code on how to efficiently get output for multiple species.
```

Bai-Li-NOAA replied:
Done. Thanks for the suggestion.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+  unique_species <- unique(data[["species"]])
+  if (!species_name %in% unique_species) {
+    cli::cli_abort(c(
+      "x" = "Species {.val {species_name}} not found in the data.",
```
```suggestion
      "x" = "Species {.val {species_name}} not found in the {.par data[['species']]}.",
```
I am not actually sure if the suggestion is technically correct but I think that it would be good for the user to be reminded which column of data is being used.

Bai-Li-NOAA replied:
Done. Thanks for the suggestion.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+
+  # Create a tibble with the labels (e.g., biomass) and types (e.g., index)
+  # Define available labels from data
+  labels <- c("biomass", "catch", "weight")
```
Should this be an internal data object? I see that an almost similar string is used in the tests, with "numbers" added.

Bai-Li-NOAA replied:
I'm thinking of leaving it as is for now since it only appears once in calc_truth.R. The similar strings in the tests are just for testing, and I don't think users need to create those labels themselves.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+  # Extract monthly data by labels
+  truth_monthly <- purrr::map(
+    labels,
+    function(x) calc_truth_monthly(data, species_name, x)
```
This is assuming that all EWE models are reported in monthly time steps. Is that true?

Bai-Li-NOAA replied:
@rklasky, does the Ecosim output in CSV format always include both monthly and yearly time step results, regardless of the time unit set in Ecopath -> Input -> Model Parameters -> Time units (e.g., year, day, or another unit)? What about Ecospace? Sorry for not testing this myself. My version of EwE is outdated and currently in read-only mode.

kellijohnson-NOAA commented on R/calc_truth.R:
```diff
@@ -0,0 +1,342 @@
+      rep("agecomp", length(labels))
+    )
+  ) |>
+    dplyr::filter(!is.na(type))
```
It was hard for me to follow what this code was doing and it seems like it will be difficult to augment it if we get future types/labels. The suggestion below also works and seems like it would be easier to maintain
```
tidyr::expand(
  # Create a tibble with zero rows but the desired columns
  dplyr::tibble("label" = character(), "type" = character()),
  # Create all combinations of label and type
  label = labels,
  type = c("agecomp", "index")
) |> 
  # Remove the weight/index combo b/c it doesn't exist
  dplyr::filter(!(label == "weight" & type == "index")) |> 
  # This line can be removed, I just added it to get the same ordering as the original tibble
  dplyr::arrange(dplyr::desc(type))
```

Bai-Li-NOAA replied:
Done, thanks for the suggestion!

kellijohnson-NOAA reviewed (approved):
I am uncertain if we should return everything as both monthly and yearly. Should we always return the most fine-grained output and then provide functions to calculate more granular output?
Can you please run styler::style_pkg() to remove the white spaces at the end of the lines?
How come "truth_" is appended to some of the column names?

@Bai-Li-NOAA here are my main notes. I love the organization of everything. I will leave it to you to filter through the comments and take what you want, e.g., if you want to keep the annual summaries in there feel free. The biggest thing that I see is that the code assumes all output from EWE is monthly but I am not sure if that is true.

Bai-Li-NOAA commented:
Thank you for your review, @kellijohnson-NOAA! I’ve addressed all your suggestions except for the one regarding time steps. I’ll check in with Ron for further input on that, and make any necessary edits afterward. Once that’s done, I’ll rebase the branch before merging it into main.

kellijohnson-NOAA commented:
Sounds great! Thank you Bai.

rklasky commented:
I'm have v. 6.7.0.18989 and I tested year and day and the output for both
was in monthly fractions of years...like:

1980
1980.083
1980.167
...

-Ron


On Mon, Aug 4, 2025 at 2:53 PM Bai Li - NOAA ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In R/calc_truth.R
> <https://github.com/NOAA-FIMS/ecosystemom/pull/5#discussion_r2252329414>:
>
> > +
> +  # Create a tibble that maps labels to their types (index or age composition)
> +  types <- dplyr::tibble(
> +    label = rep(labels, 2),
> +    type = c(
> +      # weight only has agecomp type, NA for weight index
> +      ifelse(labels %in% c("biomass", "catch"), "index", NA_character_),
> +      rep("agecomp", length(labels))
> +    )
> +  ) |>
> +    dplyr::filter(!is.na(type))
> +
> +  # Extract monthly data by labels
> +  truth_monthly <- purrr::map(
> +    labels,
> +    function(x) calc_truth_monthly(data, species_name, x)
>
> @rklasky <https://github.com/rklasky>, does the Ecosim output in CSV
> format always include both monthly and yearly time step results, regardless
> of the time unit set in Ecopath -> Input -> Model Parameters -> Time units
> (e.g., year, day, or another unit)? What about Ecospace? Sorry for not
> testing this myself. My version of EwE is outdated and currently in
> read-only mode.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-FIMS/ecosystemom/pull/5#discussion_r2252329414>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZ6RSFIX4Z4N5YFE5IJPI33L6T3FAVCNFSM6AAAAAB7IXUVWOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAOBVGM4TOMBVHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Ron Klasky
Contractor (remote)
T and T Consulting Services, Inc.
In Support of Office of Science and Technology | Ecosystem Management
NOAA | National Marine Fisheries Service
(321) 765-4465 (o)
(661) 857-4720 (c)

Bai-Li-NOAA commented:
@rklasky, Thank you for testing!
